sql-statements: document BACKUP, RESTORE and SHOW [BACKUPS|RESTORES] (#2618)
pingcap/docs: sql-statements/sql-statement-restore.md

+------------------------------+-----------+----------+---------------------+---------------------+
1 row in set (28.961 sec)
```

This will restore everything from a backup archive at the local filesystem. The data will be read as SST files from the `/mnt/backup/2020/04/` folders distributed among all TiDB and TiKV nodes.

On complete, it will display the result in one row of

| Column | Meaning |
|--------|---------|
| `Destination` | The destination URL to read from |
| `Size` |  The total size of the backup archive, in bytes |
| `BackupTS` | (not used) |
| `Queue Time` | The timestamp (in current time zone) when the `RESTORE` task was queued. |
| `Execution Time` | The timestamp (in current time zone) when the `RESTORE` task starts to run. |

### Partial restore

You can specify which databases or tables to restore. If some databases or tables are missing from the backup archive, they will be ignored, and thus `RESTORE` would complete without doing anything.

{{< copyable "sql" >}}

```sql
RESTORE DATABASE `test` FROM 'local:///mnt/backup/2020/04/';

RESTORE TABLE `test`.`sbtest01`, `test`.`sbtest02` FROM 'local:///mnt/backup/2020/04/';
```

### Remote destinations

> BR supports restoring from S3 or GCS:

Reviewer: ```suggestion
BR supports restoring data from S3 or GCS:
```

When running on cloud environment where credentials should not be distributed, set the `SEND_CREDENTIALS_TO_TIKV` option to `FALSE`:

{{< copyable "sql" >}}

```sql
RESTORE DATABASE * FROM 's3://example-bucket-2020/backup-05/?region=us-west-2'
    SEND_CREDENTIALS_TO_TIKV = FALSE;
```

### Performance fine-tuning

Use `RATE_LIMIT` to limit the average download speed per TiKV node to reduce network bandwidth.

By default, TiDB node would run 128 restore threads. This value can be adjusted with the `CONCURRENCY` option.

Before restore is completed, `RESTORE` would perform a checksum against the data from the archive to verify correctness. This step can be disabled with the `CHECKSUM` option if you are confident that this is unnecessary.

{{< copyable "sql" >}}

```sql
RESTORE DATABASE * FROM 's3://example-bucket-2020/backup-06/'
    RATE_LIMIT = 120 MB/SECOND
    CONCURRENCY = 64
    CHECKSUM = FALSE;
```

### Incremental restore

There is no special syntax to perform incremental restore. TiDB will recognize whether the backup archive is full or incremental and take appropriate action. You only need to apply each incremental restore in correct order.

> For instance, if we create backup as

Reviewer: ```suggestion
For instance, if a backup task is created as follows:
```

{{< copyable "sql" >}}

```sql
BACKUP DATABASE `test` TO 's3://example-bucket/full-backup'  SNAPSHOT = 413612900352000;
BACKUP DATABASE `test` TO 's3://example-bucket/inc-backup-1' SNAPSHOT = 414971854848000 LAST_BACKUP = 413612900352000;
BACKUP DATABASE `test` TO 's3://example-bucket/inc-backup-2' SNAPSHOT = 416353458585600 LAST_BACKUP = 414971854848000;
```

> then we should apply the restore in the same order

Reviewer: ```suggestion
then the same order should be applied in the restore:
```

---
title: RESTORE | TiDB SQL Statement Reference
summary: An overview of the usage of RESTORE for the TiDB database.
category: reference

> aliases: ['/docs/dev/reference/sql/statements/restore/']

Reviewer 2: Ditto.

---
title: RESTORE | TiDB SQL Statement Reference
summary: An overview of the usage of RESTORE for the TiDB database.
category: reference
---

# RESTORE

This statement performs a distributed restore from a backup archive previously produced by a [`BACKUP` statement](/sql-statements/sql-statement-backup.md).

The `RESTORE` statement uses the same engine as the [BR tool](/br/backup-and-restore-use-cases.md), except that the restore process is driven by TiDB itself rather than a separate BR tool. All benefits and caveats of BR also apply here. In particular, **`RESTORE` is currently not ACID-compliant**. Before running `RESTORE`, ensure that the following requirements are met:

* The cluster is "offline", and the current TiDB session is the only active SQL connection to access all tables being restored.
* When a full restore is being performed, the tables being restored should not already exist, because existing data might be overridden and causes inconsistency between the data and indices.
* When an incremental restore is being performed, the tables should be at the exact same state as the `LAST_BACKUP` timestamp when the backup is created.

> Running `RESTORE` requires `SUPER` privilege. Additionally, both the TiDB node executing the backup and all TiKV nodes in the cluster must have read permission from the destination.
> 
> The `RESTORE` statement is blocking, and will finish only after the entire backup task is finished, failed, or canceled. A long-lasting connection should be prepared for running `RESTORE`. The task can be canceled using the [`KILL TIDB QUERY`](/sql-statements/sql-statement-kill.md) statement.

Reviewer 3: ```suggestion
Running `RESTORE` requires `SUPER` privilege. Additionally, both the TiDB node executing the restore and all TiKV nodes in the cluster must have read permission from the destination.

The `RESTORE` statement is blocking, and will finish only after the entire restore task is finished, failed, or canceled. A long-lasting connection should be prepared for running `RESTORE`. The task can be canceled using the [`KILL TIDB QUERY`](/sql-statements/sql-statement-kill.md) statement.
```
@kennytm PTAL. Not sure if they're typos.

Author: included in #4342.
